### ISSUE TITLE
[BUG] mlflow-version-cast-bug

### ISSUE THREAD

**Issue by fbarulli** [mlflow/mlflow#24515]
> [!WARNING]
> Before submitting a PR, please make sure that:
> - A maintainer has triaged this issue and applied the `ready` label
> - This issue has no assignee
> - No duplicate PR exists
>
> PRs not meeting these requirements may be automatically closed.

### Issues Policy acknowledgement

- [x] I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

### Where did you encounter this bug?

Local machine

### MLflow version

- Client: 3.14.0
- Tracking server: 3.14.0

### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
- **Python version**:

### Describe the problem

MLflow 3.14.0 fails to retrieve registered model versions from a PostgreSQL backend when using the REST Model Registry API.
The SQLAlchemy query binds the version parameter as VARCHAR, while the model_versions.version column is INTEGER. PostgreSQL rejects the comparison (integer = character varying).
This only happens through the REST API path (client → server). Direct ORM usage works fine.

### Tracking information

```shell
fbarulli/mlflow-version-cast-bug
```

### Code to reproduce issue

```
fbarulli/mlflow-version-cast-bug
```

### Stack trace

```
fbarulli/mlflow-version-cast-bug
```

### Other info / logs

```
fbarulli/mlflow-version-cast-bug
```

### What component(s) does this bug affect?

- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [x] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow model serving, deployment tools, Spark UDFs
- [ ] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/prompt`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting
- [ ] `area/docs`: MLflow documentation pages

**Comment by B-Step62** [comment]
@fbarulli Can you provide a code for reproducing the issue?

**Comment by qubeena07** [comment]
I put together a fix for this at #24522 (currently closed automatically since this issue is not yet labeled ready). The root cause is in mlflow/utils/validation.py, the function that validates a model version converts it to an integer but never returns that value, so callers keep using whatever type was originally passed in. On PostgreSQL this leads to the VARCHAR versus INTEGER comparison failure described above, since the version column is strictly typed. SQLite tolerates the mismatch silently, which is why this likely was not caught earlier.

Whenever a maintainer gets a chance to triage this and add the ready label, I would appreciate it if the PR could be reopened.

**Comment by B-Step62** [comment]
@qubeena07 Did you reproduce the issue?

**Comment by fbarulli** [comment]
@B-Step62 yes ofc, its [here](https://github.com/fbarulli/mlflow-version-cast-bug) and also:



## Steps to reproduce

### 1. Start MLflow server with PostgreSQL backend

    mlflow server \
      --host 0.0.0.0 \
      --backend-store-uri postgresql+psycopg://mlflow:mlflow@postgres:5432/mlflow

### 2. Register a model

Example:

    import mlflow
    from sklearn.datasets import load_diabetes
    from sklearn.ensemble import RandomForestRegressor

    mlflow.set_tracking_uri("http://mlflow:5000")

    mlflow.set_experiment("test")

    with mlflow.start_run():

        X, y = load_diabetes(return_X_y=True)

        model = RandomForestRegressor()
        model.fit(X, y)

        mlflow.sklearn.log_model(
            model,
            name="model",
            registered_model_name="demo-model",
        )

### 3. Retrieve model version

    import mlflow

    client = mlflow.MlflowClient()

    client.get_model_version(
        "demo-model",
        3,
    )

---

## Expected behavior

A ModelVersion object should be returned.

Example:

    ModelVersion(
        name="demo-model",
        version="3",
        status="READY"
    )

---

## Actual behavior

The request fails with:

    mlflow.exceptions.RestException:
    BAD_REQUEST:
    (psycopg.errors.UndefinedFunction)
    operator does not exist:
    integer = character varying

---

## Generated SQL

The failing query:

    SELECT model_versions.workspace,
           model_versions.name,
           model_versions.version
    FROM model_versions
    WHERE model_versions.workspace = %(workspace_1)s::VARCHAR
      AND model_versions.name = %(name_1)s::VARCHAR
      AND model_versions.version = %(version_1)s::VARCHAR
      AND model_versions.current_stage != %(current_stage_1)s::VARCHAR

Parameters:

    {
        "workspace_1": "default",
        "name_1": "demo-model",
        "version_1": "3",
        "current_stage_1": "Deleted_Internal"
    }

The problematic comparison:

    model_versions.version = '3'::VARCHAR

---

## PostgreSQL error

    psycopg.errors.UndefinedFunction:

    operator does not exist:
    integer = character varying

PostgreSQL reports:

    No operator matches the given name and argument types.
    You might need to add explicit type casts.

---

## MLflow source location

The issue appears in:

    mlflow/store/model_registry/sqlalchemy_store.py

Function:

    _get_sql_model_version()

Current condition:

    conditions = [
        SqlModelVersion.name == name,
        SqlModelVersion.version == version,
        SqlModelVersion.current_stage != STAGE_DELETED_INTERNAL,
    ]

The version value arriving through the REST API path is treated as a string.

---

## Additional investigation

The ORM model defines:

    version = Column(Integer, nullable=False)

Direct SQLAlchemy usage works:

    session.query(SqlModelVersion).filter(
        SqlModelVersion.name == "demo-model",
        SqlModelVersion.version == 1,
    ).all()

This generates:

    model_versions.version = 1

and succeeds.

The failure only occurs through the MLflow REST registry API path.

---

## Possible fix

Explicitly bind the parameter as an integer:

    from sqlalchemy import bindparam, Integer

    conditions = [
        SqlModelVersion.name == name,
        SqlModelVersion.version == bindparam(
            "version",
            value=int(version),
            type_=Integer(),
        ),
        SqlModelVersion.current_stage != STAGE_DELETED_INTERNAL,
    ]

This ensures SQLAlchemy sends an INTEGER bind parameter instead of VARCHAR.

---

## Impact

Affected operations include:

- MlflowClient.get_model_version()
- Model registry operations that internally retrieve a specific model version

This prevents retrieving registered model versions when using PostgreSQL as the MLflow backend store.

---

## Minimal reproducer

Repository:

    mlflow-alias-bug/
    ├── Dockerfile
    ├── docker-compose.yml
    ├── requirements.txt
    ├── reproduce.py
    └── README.md

Dependencies:

    mlflow==3.14.0
    SQLAlchemy==2.0.51
    psycopg==3.3.4
    PostgreSQL 16

**Comment by qubeena07** [comment]
Thanks for the repro, @fbarulli that confirms the root cause.

@B-Step62 fix is ready at #24522 (validator now returns the converted int instead of discarding it, call sites updated in both the SqlAlchemy store and file store). It's currently auto-closed pending the ready label on this issue, could you add it so the PR can be reopened?

**Comment by fbarulli** [comment]
@qubeena07, np, ive added the fix in the repo, as it happens in another function as well.



should i leave my cv somewhere? lol.

**Comment by qubeena07** [comment]
@fbarulli Haha I would ask for your resume, but you're already doing such a great job working for free.

**Comment by B-Step62** [comment]
Thanks both!

@qubeena07 Attached ready label, feel free to reopen the PR.

**Comment by aaronteo-db** [comment]
Thanks for your contribution @fbarulli and @qubeena07 ! We have merged the fix PR - https://github.com/mlflow/mlflow/pull/24565
